types/react-transition-group/Transition.d.ts

```diff
@@ -1,8 +1,27 @@
-import { Component, ReactNode } from "react";
-
-export type EndHandler = (node: HTMLElement, done: () => void) => void;
-export type EnterHandler = (node: HTMLElement, isAppearing: boolean) => void;
-export type ExitHandler = (node: HTMLElement) => void;
+import { Component, ReactNode } from 'react';
+
+type RefHandler<
+    RefElement extends undefined | HTMLElement,
+    ImplicitRefHandler extends (node: HTMLElement, ...args: any[]) => void,
+    ExplicitRefHandler extends (...args: any[]) => void
+> = {
+    implicit: ImplicitRefHandler;
+    explicit: ExplicitRefHandler;
+}[RefElement extends undefined ? 'implicit' : 'explicit'];
+
+export type EndHandler<RefElement extends undefined | HTMLElement> = RefHandler<
+    RefElement,
+    (node: HTMLElement, done: () => void) => void,
+    (done: () => void) => void
+>;
+
+export type EnterHandler<RefElement extends undefined | HTMLElement> = RefHandler<
+    RefElement,
+    (node: HTMLElement, isAppearing: boolean) => void,
+    (isAppearing: boolean) => void
```

Comment:
FYI. If the PR https://github.com/reactjs/react-transition-group/pull/469 is merged, it would need to update the definition.

Reply:
In case https://github.com/reactjs/react-transition-group/pull/469 is merged a lot of the stuff in this PR will be obsolute, and can be removed (the change will be a single line one - adding the `nodeRef` prop). At any rate, it would be introduced in a new version of `react-transition-group` (and a breaking one, see https://github.com/reactjs/react-transition-group/pull/469#issue-260515743), so we can always do that in `@types/react-transition-group@5.0.0` whenever that's merged. In the meantime though, 4.4.0 has no type definitions that are properly typed, so I think this PR is still relevant.

Reply:
Yes, you're right. 👍